superset/commands/database/resync_permissions.py

```diff
@@ -0,0 +1,276 @@
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+from functools import partial
+from typing import Iterable
+
+from superset import security_manager
+from superset.commands.base import BaseCommand
+from superset.commands.database.exceptions import (
+    DatabaseConnectionFailedError,
+    DatabaseConnectionResyncPermissionsError,
+    DatabaseNotFoundError,
+)
+from superset.commands.database.utils import ping
+from superset.daos.database import DatabaseDAO
+from superset.daos.dataset import DatasetDAO
+from superset.databases.ssh_tunnel.models import SSHTunnel
+from superset.db_engine_specs.base import GenericDBException
+from superset.exceptions import OAuth2RedirectError
+from superset.models.core import Database
+from superset.utils.decorators import on_error, transaction
+
+logger = logging.getLogger(__name__)
+
+
+class ResyncPermissionsCommand(BaseCommand):
+    """
+    Command to resync database permissions.
+    """
+
+    def __init__(
+        self,
+        model_id: int,
+        old_db_connection_name: str | None = None,
+        db_connection: Database | None = None,
+        ssh_tunnel: SSHTunnel | None = None,
+    ):
+        """
+        Constructor method.
+        """
+        self.db_connection_id = model_id
+        self.old_db_connection_name: str | None = old_db_connection_name
+        self.db_connection: Database | None = db_connection
+        self.db_connection_ssh_tunnel: SSHTunnel | None = ssh_tunnel
+
+    def validate(self) -> None:
+        if not self.db_connection:
+            database = DatabaseDAO.find_by_id(self.db_connection_id)
+            if not database:
+                raise DatabaseNotFoundError()
+            self.db_connection = database
+
+        if not self.old_db_connection_name:
+            self.old_db_connection_name = self.db_connection.database_name
+
+        if not self.db_connection_ssh_tunnel:
+            self.db_connection_ssh_tunnel = DatabaseDAO.get_ssh_tunnel(
+                self.db_connection_id
+            )
+
+        with self.db_connection.get_sqla_engine() as engine:
+            try:
+                alive = ping(engine)
+            except Exception as err:
+                raise DatabaseConnectionFailedError() from err
+
+        if not alive:
+            raise DatabaseConnectionFailedError()
+
+    @transaction(
+        on_error=partial(on_error, reraise=DatabaseConnectionResyncPermissionsError)
+    )
+    def run(self) -> None:
```

Comment:
This logic was moved from `superset/commands/database/update.py`.

Reply:
I had to pass `old_db_connection_name` and `db_connection` around (even though they're available through `self`) because `mypy` was not happy using something that technically could be `None` (even though that's validated in the `validate` method).

Reply:
You can probably do something like:

```python
self.validate()  # sets self._model or raises
self._model = cast(Database, self._model)
```

Reply:
@betodealmeida doing the `cast()` wouldn't fully solve the issue, unless I do the `cast` at all helper methods. I've decided to keep `self._old_db_connection_name` as `str | None` and `self._db_connection` as `Database | None` and then create `self.db_connection: Database` and `self.old_db_connection_name: str` at the `validate` method, and use these at all later methods.